### PR TITLE
decouple sessions and client

### PR DESCRIPTION
**Pull request by braelyn-ai**
Enables a single client to have multiple concurrent sessions.
Adds support for REST servers.

### CODE REVIEW

File: agentops/helpers.py [hunk 1]
```diff
@@ -30,6 +30,21 @@ def getinstance(*args, **kwargs):
     return getinstance
 
 
+def conditional_singleton(class_):
```
Comment (mbarnathan) [inline agentops/helpers.py:34]: Had to do something similar in #257, in case a conflict emerges

File: agentops/__init__.py [hunk 2]
```diff
@@ -79,14 +81,20 @@ def init(
...
-    return inherited_session_id or c.current_session_id
+    if inherited_session_id:
+        return inherited_session_id
+    elif len(c.current_session_ids) is 1:
+        return c.current_session_ids[0]
+    else:
+        return None
 
 
 def end_session(
```
Comment (mbarnathan) [inline agentops/__init__.py:111]: Not saying this has to be done here, but a session context manager could be a cool future addition

File: agentops/__init__.py [hunk 3]
```diff
@@ -161,18 +176,17 @@ def record(event: Union[Event, ErrorEvent]):
...
 
 
-@check_init
 def add_tags(tags: List[str]):
     """
     Append to session tags at runtime.
 
     Args:
         tags (List[str]): The list of tags to append.
+        session_id (str, optional): which session to add tags to if using multiple concurrent sessions
```
Comment (siyangqiu) [inline agentops/__init__.py]: I don't see session_id as an argument. Is that intended?

File: agentops/__init__.py [hunk 4]
```diff
@@ -89,24 +90,30 @@ def init(
         max_wait_time=max_wait_time,
         max_queue_size=max_queue_size,
         tags=tags,
-        override=override,
         instrument_llm_calls=instrument_llm_calls,
-        auto_start_session=auto_start_session,
+        auto_start_session=False,
```
Comment (siyangqiu) [inline agentops/__init__.py]: Is this intended or left over from testing?

File: agentops/session.py [hunk 5]
```diff
@@ -45,25 +68,185 @@ def set_session_video(self, video: str) -> None:
...
+
+    def set_tags(self, tags):
+        if not (isinstance(tags, list) and all(isinstance(item, str) for item in tags)):
+            if isinstance(tags, str):  # if it's a single string
+                tags = [tags]  # make it a list
+
+        self.tags = tags
+        self._update_session()
+
+    def record(potato, event: Union[Event, ErrorEvent]):
```
Comment (siyangqiu) [inline agentops/session.py]: potato????

File: agentops/session.py [hunk 6]
```diff
@@ -45,25 +68,185 @@ def set_session_video(self, video: str) -> None:
...
+                    == event.trigger_event.end_timestamp
+                ):
+                    event.trigger_event.end_timestamp = get_ISO_time()
+
+                event.trigger_event_id = event.trigger_event.id
+                event.trigger_event_type = event.trigger_event.event_type
+                potato.record(event)
+                event.trigger_event = None  # removes trigger_event from serialization
+
+        potato._add_event(event.__dict__)
```
Comment (siyangqiu) [inline agentops/session.py]: ...

File: agentops/helpers.py [hunk 7]
```diff
@@ -30,6 +31,25 @@ def getinstance(*args, **kwargs):
...
+                instances[class_] = class_(*args, **kwargs)
+            return instances[class_]
+        else:
+            return class_(*args, **kwargs)
+
+    return getinstance
+
+
+def clear_singletons():
+    global instances
```
Comment (siyangqiu) [inline agentops/helpers.py]: Do we want to want to prefix this with ao or agentops?

File: agentops/exceptions.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class MultiSessionException(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+
+
+class NoSessionException(Exception):
+    def __init__(self, message):
+        super().__init__(message)
```
Comment (siyangqiu) [inline agentops/exceptions.py:11]: move the configuration exception here as well

File: agentops/client.py [hunk 9]
```diff
@@ -19,24 +19,27 @@
...
 from .helpers import (
     get_ISO_time,
-    singleton,
     check_call_stack_for_agent_id,
     get_partner_frameworks,
+    singleton,
+    conditional_singleton,
+    safe_serialize,
 )
+from .http_client import HttpClient
```
Comment (siyangqiu) [inline agentops/client.py]: I don't this is being used in client

**Review by siyangqiu — APPROVED** [review]
Small changes

File: agentops/__init__.py (hunk 3 above)
Comment (braelyn-ai) [reply agentops/__init__.py]: unintended! will remove!

File: agentops/session.py (hunk 5 above)
Comment (braelyn-ai) [reply agentops/session.py]: potato

File: agentops/session.py (hunk 6 above)
Comment (braelyn-ai) [reply agentops/session.py]: 🥔

File: agentops/__init__.py [hunk 10]
```diff
@@ -65,11 +66,10 @@ def init(
         max_queue_size (int, optional): The maximum size of the event queue. Defaults to 100.
         tags (List[str], optional): Tags for the sessions that can be used for grouping or
             sorting later (e.g. ["GPT-4"]).
-        override (bool, optional): [Deprecated] Use `instrument_llm_calls` instead. Whether to instrument LLM calls and emit LLMEvents..
-        instrument_llm_calls (bool): Whether to instrument LLM calls and emit LLMEvents..
+        instrument_llm_calls (bool): Whether to instrument LLM calls and emit LLMEvents.
         auto_start_session (bool): Whether to start a session automatically when the client is created.
         inherited_session_id (optional, str): Init Agentops with an existing Session
-        skip_auto_end_session (optional, bool): Don't automatically end session based on your framework's decision making
+        skip_auto_end_session (optional, bool): Don't automatically end session based on your framework's decision-making
```
Comment (HowieG) [inline agentops/__init__.py]: Wording here could be clearer

File: agentops/__init__.py [hunk 11]
```diff
@@ -189,7 +199,7 @@ def get_api_key() -> str:
 
 def set_parent_key(parent_key):
     """
-    Set the parent API key which has visibility to projects it is parent to.
+    Set the parent API key which has visibility to projects it is parented to.
```
Comment (HowieG) [inline agentops/__init__.py]: parent

File: agentops/agent.py [hunk 12]
```diff
@@ -18,8 +18,15 @@ def new_init(self, *args, **kwargs):
                 try:
                     original_init(self, *args, **kwargs)
                     self.agent_ops_agent_id = str(uuid4())
+
+                    session = kwargs.get("session", None)
```
Comment (HowieG) [inline agentops/agent.py:22]: Let's not do session as a parameter to LLM calls. More work for us bc we'll have to do this to every LLM we support

File: agentops/agent.py [hunk 13]
```diff
@@ -18,8 +18,15 @@ def new_init(self, *args, **kwargs):
...
+                    if session is not None:
+                        self.agent_ops_session_id = session.session_id
```
Comment (HowieG) [inline agentops/agent.py:24]: agentops not agent_ops

Comment (HowieG) [reply agentops/agent.py:24]: actually agent_session_id or assigned_to_session or something that makes it clearer that it's not assigned to all of agentops but to an agent specifically

File: agentops/config.py [hunk 14]
```diff
@@ -1,16 +1,18 @@
 """
-AgentOps configuration.
+AgentOps Client configuration.
 
 Classes:
-    Configuration: Stores the configuration settings for AgentOps clients.
+    ClientConfiguration: Stores the configuration settings for AgentOps clients.
+    AgentOpsConfiguration: This is a class to store the configuration settings for initializing AgentOps
```
Comment (HowieG) [inline agentops/config.py]: Remove

File: agentops/decorators.py [hunk 15]
```diff
@@ -18,17 +21,19 @@ def decorator(func):
...
+                    func, event_name, *args, session=session, **kwargs
                 )
 
             return async_wrapper
         else:
 
             @functools.wraps(func)
-            def sync_wrapper(*args, **kwargs):
-                return Client()._record_event_sync(func, event_name, *args, **kwargs)
+            def sync_wrapper(*args, session: Optional[Session] = None, **kwargs):
```
Comment (HowieG) [inline agentops/decorators.py:33]: move session to end

File: agentops/decorators.py [hunk 16]
```diff
@@ -18,17 +21,19 @@ def decorator(func):
         if inspect.iscoroutinefunction(func):
 
             @functools.wraps(func)
-            async def async_wrapper(*args, **kwargs):
+            async def async_wrapper(*args, session: Optional[Session] = None, **kwargs):
```
Comment (HowieG) [inline agentops/decorators.py:24]: move session to end

File: agentops/exceptions.py [hunk 17]
```diff
@@ -0,0 +1,19 @@
+from .log_config import logger
+
+
+class MultiSessionException(Exception):
+    def __init__(self, message):
+        super().__init__(message)
```
Comment (HowieG) [inline agentops/exceptions.py:6]: inconsistency here between MultiSessionException and ConfigurationError
The latter does logger.warning(message). I think we need to do that here. I did it this way so the catcher doesn't have to remember to call logger.warning it's done for them

File: agentops/helpers.py (hunk 1 above)
Comment (HowieG) [reply agentops/helpers.py:34]: I think we should remove if we're not using

File: agentops/helpers.py [hunk 18]
```diff
@@ -18,18 +18,38 @@
...
+                ao_instances[class_] = class_(*args, **kwargs)
+            return ao_instances[class_]
+        else:
+            return class_(*args, **kwargs)
 
     return getinstance
 
 
+def clear_singletons():
+    global ao_instances
```
Comment (HowieG) [inline agentops/helpers.py:49]: Can we check if this is where global goes or if it should be on line 21 where we define ao_instances?

File: agentops/llm_tracker.py [hunk 19]
```diff
@@ -348,9 +403,14 @@ def handle_stream_chunk(chunk):
                     pass
 
             except Exception as e:
-                self.client.record(
-                    ErrorEvent(trigger_event=self.llm_event, exception=e)
-                )
+                if session is not None:
```
Comment (HowieG) [inline agentops/llm_tracker.py]: can just do if session: to stay consistent

File: agentops/llm_tracker.py [hunk 20]
```diff
@@ -407,9 +467,17 @@ def generator():
             self.llm_event.completion_tokens = response.meta.tokens.output_tokens
             self.llm_event.model = kwargs.get("model", "command-r-plus")
 
-            self.client.record(self.llm_event)
+            if session is not None:
```
Comment (HowieG) [inline agentops/llm_tracker.py]: can just do if session: to stay consistent

File: agentops/llm_tracker.py [hunk 21]
```diff
@@ -470,9 +543,14 @@ def override_openai_v1_completion(self):
 
         def patched_function(*args, **kwargs):
             init_timestamp = get_ISO_time()
+            session = kwargs.get("session", None)
```
Comment (HowieG) [inline agentops/llm_tracker.py:499]: remove

**Review by HowieG — APPROVED** [review]

File: agentops/agent.py (hunk 13 above)
Comment (braelyn-ai) [reply agentops/agent.py:24]: crew accesses these properties so unfortunately changing these would require a crew update

File: agentops/decorators.py (hunk 15 above)
Comment (braelyn-ai) [reply agentops/decorators.py:33]: **kwargs has to be at the end

File: agentops/helpers.py (hunk 18 above)
Comment (braelyn-ai) [reply agentops/helpers.py:49]: you put global before its use

File: agentops/llm_tracker.py (hunk 20 above)
Comment (braelyn-ai) [reply agentops/llm_tracker.py]: added helper function

**Pull request merged**